sled-hardware/src/disk.rs

```diff
@@ -341,23 +404,91 @@ impl Disk {
 
     // Ensure that the zpool contains all the datasets we would like it to
     // contain.
-    fn ensure_zpool_has_datasets(
+    async fn ensure_zpool_has_datasets(
+        log: &Logger,
         zpool_name: &ZpoolName,
+        disk_identity: &DiskIdentity,
+        key_requester: Option<&StorageKeyRequester>,
     ) -> Result<(), DiskError> {
-        let datasets = match zpool_name.kind().into() {
-            DiskVariant::M2 => M2_EXPECTED_DATASETS.iter(),
-            DiskVariant::U2 => U2_EXPECTED_DATASETS.iter(),
+        let (root, datasets) = match zpool_name.kind().into() {
+            DiskVariant::M2 => (None, M2_EXPECTED_DATASETS.iter()),
+            DiskVariant::U2 => {
+                (Some(CRYPT_DATASET), U2_EXPECTED_DATASETS.iter())
+            }
         };
-        for dataset in datasets.into_iter() {
+
+        let zoned = false;
+        let do_format = true;
+
+        // Ensure the root encrypted filesystem exists
+        // Datasets below this in the hierarchy will inherit encryption
+        if let Some(dataset) = root {
+            let Some(key_requester) = key_requester else {
+                return Err(DiskError::MissingStorageKeyRequester);
+            };
             let mountpoint = zpool_name.dataset_mountpoint(dataset);
+            let keypath: Keypath = disk_identity.into();
+
+            let epoch =
+                if let Ok(epoch_str) = Zfs::get_oxide_value(dataset, "epoch") {
+                    if let Ok(epoch) = epoch_str.parse::<u64>() {
+                        epoch
+                    } else {
+                        return Err(DiskError::MissingEpochProperty(
+                            dataset.to_string(),
+                        ));
```

Comment:
This error seems a little misleading - the property exists, it just can't be parsed.